File: 01-programming/csharp/concepts/code/01-basics/syntax-init.cs
// -o-o-o-o- Syntax (Initialisation) -o-o-o-o-

/* This file goes over variable declaration and initialisation in C# simply.
 *
 *  # # # # KEY # # # #
 *
 * declaration     - creating a variable and giving it a name and type.
 *                   eg. int a;
 *                   Same concept as C.
 *
 * initialisation  - assigning a value to a variable at the time of declaration.
 *                   eg. int a = 10;
 *                   Same concept as C.
 *
 * = - the assignment operator. Stores a value in a variable.
 *     Same as in C.
 *
 * Data types in C# (compared to C):
 *
 *   int     - 32-bit integer.                    Same as int in C.
 *   long    - 64-bit integer.                    Same as long in C.
 *   float   - 32-bit decimal. Needs 'f' suffix.  Same as float in C.
 *   double  - 64-bit decimal.                    Same as double in C.
 *   char    - a single character.                Same as char in C.
 *   bool    - true or false. Written as bool.    Same idea as in C (C uses int).
 *   string  - a sequence of characters.          C# has a proper string type;
 *                                                C uses char arrays.
 *   decimal - 128-bit high-precision decimal.    No equivalent in C.
 *
 * var     - lets C# infer the type automatically based on the assigned value.
 *           eg. var x = 10; is treated as int x = 10;
 *           No equivalent in C.
 *
 * default values - unassigned class-level variables get a default value:
 *                  int = 0, double = 0.0, bool = false, string = null.
 *                  Local variables MUST be assigned before use in C#.
 *
 * Console.WriteLine() - prints text to the console with a newline at the end.
 *
 * */

using System;

class Program {
    static void Main() {

        // Declaration and initialisation of basic types
        int     myInt     = 42;
        long    myLong    = 1234567890123L; // 'L' suffix for long literals
        float   myFloat   = 3.14f;          // 'f' suffix required for float literals
        double  myDouble  = 3.14159265;
        char    myChar    = 'A';
        bool    myBool    = true;
        string  myString  = "Hello";
        decimal myDecimal = 99.99m;         // 'm' suffix for decimal literals

        // Printing each variable
        Console.WriteLine($"int:     {myInt}");
        Console.WriteLine($"long:    {myLong}");
        Console.WriteLine($"float:   {myFloat}");
        Console.WriteLine($"double:  {myDouble}");
        Console.WriteLine($"char:    {myChar}");
        Console.WriteLine($"bool:    {myBool}");
        Console.WriteLine($"string:  {myString}");
        Console.WriteLine($"decimal: {myDecimal}");

        // Using 'var' — type is inferred by the compiler
        var inferredInt    = 10;       // compiler knows this is int
        var inferredString = "World";  // compiler knows this is string
        Console.WriteLine($"\nvar int:    {inferredInt}");
        Console.WriteLine($"var string: {inferredString}");

        // Declaration without initialisation (must assign before use)
        int uninitialised;
        uninitialised = 100; // assigned before it is used below
        Console.WriteLine($"\nLate initialisation: {uninitialised}");

    }
}

File: 01-programming/python/concepts/code/01-basics/syntax-init.py
# -o-o-o-o- Syntax (Initialisation) -o-o-o-o-

# This file goes over variable declaration and initialisation in Python simply.
#
#  # # # # KEY # # # #
#
# In Python, there is no separate declaration step. A variable is created
# the moment you assign a value to it. You never write the type.
#
# C:  int a;       <- declaration (no value yet)
#     a = 10;      <- initialisation
# C#: int a = 10;  <- declaration + initialisation together
# Py: a = 10       <- just assignment; Python handles everything else
#
# Data types Python uses automatically:
#
#   int     - whole numbers, any size.      eg. 42, -7, 1000000
#   float   - decimal numbers.              eg. 3.14, -0.5
#   complex - complex numbers.              eg. 2 + 3j   (no C/C# equivalent)
#   bool    - True or False (capitalised).  eg. True, False
#   str     - text / strings.               eg. "hello", 'world'
#   NoneType- the absence of a value.       eg. None  (like null in C#, NULL in C)
#
# type()   - returns the type of a variable. Useful for inspection.
#            eg. type(42) returns <class 'int'>
#
# int()    - converts a value to an integer.  eg. int("42") -> 42
# float()  - converts a value to a float.     eg. float("3.14") -> 3.14
# str()    - converts a value to a string.    eg. str(100) -> "100"
# bool()   - converts a value to a boolean.   eg. bool(0) -> False
#
# print()  - prints text to the console with a newline at the end.
# f"..."   - f-string. Embeds variables directly in a string.
#

# Basic type initialisation
my_int     = 42
my_float   = 3.14159265
my_bool    = True
my_string  = "Hello"
my_none    = None          # represents the absence of a value (like null)
my_complex = 2 + 3j        # complex number (unique to Python)

# Printing each variable and its type
print(f"int:     {my_int}      -> {type(my_int)}")
print(f"float:   {my_float} -> {type(my_float)}")
print(f"bool:    {my_bool}     -> {type(my_bool)}")
print(f"str:     {my_string}   -> {type(my_string)}")
print(f"None:    {my_none}     -> {type(my_none)}")
print(f"complex: {my_complex}   -> {type(my_complex)}")

# Type conversion (casting)
print("\nType conversion:")
print(f"int('99'):     {int('99')}    -> {type(int('99'))}")
print(f"float('3.5'):  {float('3.5')} -> {type(float('3.5'))}")
print(f"str(100):      {str(100)}     -> {type(str(100))}")
print(f"bool(0):       {bool(0)}  -> {type(bool(0))}")
print(f"bool(1):       {bool(1)}  -> {type(bool(1))}")

# Multiple assignment on one line
x, y, z = 10, 20.5, "Python"
print(f"\nx={x}, y={y}, z={z}")

# Assign the same value to multiple variables
a = b = c = 0
print(f"a={a}, b={b}, c={c}")
